From 4f8584543ea25a01c178b2a1b2812fca357f03b3 Mon Sep 17 00:00:00 2001
From: Chen Zhongjin <chenzhongjin@huawei.com>
Date: Tue, 28 Mar 2023 11:50:40 -0400
Subject: [PATCH 646/921] [Huawei] bpf: Check sock lock when updating sock
 status

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 188591

--------------------------------

bpf_xdp_update_tcp_seq() update tcp status of socket with checking
socket by sock_owned_by_user(sk). And this triggered warning:

WARNING: CPU: 5 PID: 0 at include/net/sock.h:1692 bpf_xdp_update_tcp_seq+0x59c/0x750
...

This means when the bpf progs are updating status, the socket can also
be processed by userspace, which is possible to raise conflict with bpf
updating.

To make bpf_xdp_update_tcp_seq() more safe, lock socket and check if
it is holded by userspace. This can also protect socket being modified
by different bpf progs at the same time.

Fixes: 7b1fe3b7833e ("[Huawei] samples: bpf: Add sample BMC for Redis")
Signed-off-by: Chen Zhongjin <chenzhongjin@huawei.com>
---
 net/core/filter.c | 18 ++++++++++++------
 1 file changed, 12 insertions(+), 6 deletions(-)

diff --git a/net/core/filter.c b/net/core/filter.c
index c2846faf1d01..75db9a5b3cff 100644
--- a/net/core/filter.c
+++ b/net/core/filter.c
@@ -6304,39 +6304,45 @@ BPF_CALL_5(bpf_xdp_update_tcp_seq, struct xdp_buff *, ctx,
 	struct sock *sk;
 	struct tcp_sock *tp;
 	u32 seq, ack;
+	int err = -1;
 
 	if (unlikely(len != sizeof(struct bpf_tcp_update_tuple)))
-		return -1;
+		return err;
 
 	if (tuple->family == AF_INET)
 		len = sizeof(tuple->sock.ipv4);
 	else if (tuple->family == AF_INET6)
 		len = sizeof(tuple->sock.ipv6);
 	else
-		return -1;
+		return err;
 
 	sk = __bpf_sk_lookup(NULL, &tuple->sock, len, caller_net,
 			     ifindex, IPPROTO_TCP, netns_id,
 			     flags);
 	if (!sk)
-		return -1;
+		return err;
 
 	tp = tcp_sk(sk);
 
 	seq = tuple->seq;
 	ack = tuple->ack_seq;
 
+	bh_lock_sock(sk);
+	if (sock_owned_by_user(sk))
+		goto out;
+
 	if (seq == tp->copied_seq && seq == tp->rcv_nxt &&
 	    ack == tp->write_seq && ack == tp->snd_nxt) {
-		sock_owned_by_me((struct sock *)tp);
 		tp->bytes_acked = tp->bytes_sent;
 		tcp_rcv_nxt_update(tp, seq, tuple->rcv_len);
 		tcp_snd_nxt_update(tp, ack, tuple->snd_len);
 
-		return 0;
+		err = 0;
 	}
 
-	return -1;
+out:
+	bh_unlock_sock(sk);
+	return err;
 }
 
 static const struct bpf_func_proto bpf_xdp_update_tcp_seq_proto = {
-- 
2.17.1

